src/module.rs

```diff
@@ -30,41 +30,96 @@ impl WasmModule {
+    ///
+    /// > The function index space begins with an index for each imported
+    /// > function, in the order the imports appear in the Import Section,
+    /// > if present, followed by an index for each function in the Function Section,
+    /// > if present, in the order of that section.
     pub fn functions(&self) -> impl Iterator<Item = WasmFunction> {
         let function_count = self.module.functions_space();
         if function_count == 0 {
             return Either::Left(iter::empty::<WasmFunction>());
         }
 
-        let bodies = self.function_bodies();
-        let types = self.function_types();
-        assert_eq!(function_count, bodies.size_hint().0);
-        assert_eq!(function_count, types.size_hint().0);
+        let imported_functions = self.imported_functions();
+        let imported_count = self.count_imported_functions();
+        let own_count = self.count_own_functions();
+        assert_eq!(function_count, imported_count + own_count);
 
-        let functions = types
-            .zip(bodies)
+        let own_functions = self.function_types()
+            .zip(self.function_bodies())
             .enumerate()
-            .map(move |(id, (ty, body))| {
-                     let name = self.get_function_name(id);
-                     let is_export = name.is_some(); // TODO(slim): Make this more robust later.
-                     WasmFunction::new(id, ty, name, body, is_export)
-                 });
-
-        Either::Right(functions)
-    }
+            .map(move |(i, (ty, body))| {
+                // Functions from the module function section appear
+                // after imported functions, in the index space.
+                let id = imported_count + i;
+                let name = self.get_function_name(id);
+                WasmFunction {
+                    id,
+                    ty,
+                    name,
+                    body: Some(body),
+                    source: SourceSection::Function,
+                }
+            });
 
-    pub fn print_functions(&self) {
-        for f in self.functions() {
-            println!("{}", f);
-        }
+        Either::Right(imported_functions.chain(own_functions))
     }
 
     fn exported_function_names(&self) -> HashMap<usize, String> {
         let mut names = HashMap::new();
         for export in self.exports() {
-            if let Internal::Function(id) = export.internal() {
-                let name = export.field().to_owned();
-                names.insert(*id as usize, name);
+            match export.internal() {
+                Internal::Function(id) => {
+                    // NOTE(slim): `id` is an index into the function index space,
+                    // not the types section or the function section.
+                    let name = export.field().to_owned();
+                    names.insert(*id as usize, name);
+                }
+                _ => {}
```

Comment:
Error handling here.

Reply:
We want to skip over exports that aren't functions. Added a comment to clarify.